add cli executable for data_split

keonkim opened the pull request:
Simple CLI wrapper around data_split

rcurtin commented on src/mlpack/methods/preprocess/preprocess_split_main.cpp:
```diff
@@ -0,0 +1,77 @@
+// Define parameters for labels
+PARAM_STRING_REQ("input_label", "File containing labels", "I");
+PARAM_STRING_REQ("output_train_label", "File name to save train label", "l");
+PARAM_STRING_REQ("output_test_label", "File name to save test label", "L");
```
I think for consistency with other programs in mlpack I would go with parameters --training_file (-t), --training_labels_file (-l), --test_file (-T), and --test_labels_file (-L) for the output parameters.

keonkim replied:
Sure, I agree. I will update it.

zoq commented on src/mlpack/methods/preprocess/preprocess_split_main.cpp:
```diff
@@ -0,0 +1,77 @@
+  data::Load(inputFile, data, true);
+  data::Load(inputLabel, labels, true);
+  arma::Row<size_t> labels_row = labels.row(0); // extract first row
+
```
Do we assume decimal labels, so we can't use the tool in combination with e.g. linear regression?

keonkim replied:
I only thought of integer labels by looking at the test data. How should I allow decimal labels?

zoq commented on src/mlpack/methods/preprocess/preprocess_split_main.cpp:
```diff
@@ -0,0 +1,77 @@
+  arma::Row<size_t> labels_row = labels.row(0); // extract first row
+
+  // Split Data
+  const auto value = data::TrainTestSplit(data, labels_row, testRatio);
```
Does make_tuple call the copy constructor? If that's the case we should avoid this interface in favor of the pass by reference interface.

keonkim replied:
I think so. As discussed in [this pull](https://github.com/mlpack/mlpack/pull/523), I think having 6 parameters for TrainTestSplit(input, train, test, and other 3 for labels) can be used as an alternative.

keonkim commented:
I added TrainTestSplit() and renamed old ones to LabelTrainTestSplit() as discussed in #651 .
This is just a naive implementation mostly copied from Tham's work.
I believe LabelTrainTestSplit can just reuse the code in TrainTestSplit twice for both data and labels.

rcurtin commented:
I don't think you should rename the existing `TrainTestSplit()` to `LabelTrainTestSplit()`, you should just leave it named `TrainTestSplit()` and add additional overloads for when labels are not present.  To be honest, I think it would also be fine if we renamed the function the simpler `Split()`, but I am not picky about that part.

keonkim commented:
I assume that the name TrainTestSplit was borrowed from [sklearn](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.train_test_split.html).
For now, I just renamed TrainTestSplit() to Split() and added newly made functions as overloads.

keonkim commented on src/mlpack/methods/preprocess/preprocess_split_main.cpp:
```diff
@@ -0,0 +1,76 @@
+  mat training = get<0>(value);
+  mat test = get<1>(value);
+  mat trainingLabels = get<2>(value);
+  mat testLabels = get<3>(value);
```
I need to cast current Mat<double> into string, but armadillo doesn't seem to provide Matrix with strings.
I am searching for the best way to do this.

stereomatchingkiss commented:
@keonkim I cannot find a good way to cut the trailing zeros either. Anyone have a good idea?If this is a critical problem must be solved(it do not sound like for me), the dumbness solution I could come up is 
iterate the value one by one, restrict the precision of the double, this solution may need to give more params to the Save api.

rcurtin commented:
The issue is that in `armadillo_bits/diskio_meat.hpp`, we have this bit of code:

```
if( (is_float<eT>::value == true) || (is_double<eT>::value == true) )
  {
  f.setf(ios::scientific);
  f.precision(12);
  cell_width = 20;
  }
```

and this gives us no way to set the stream how we like.  Reimplementing `save_raw_ascii()` and calling our own overload is probably not a great idea because `ios::scientific` is the best way to represent arbitrary numbers...

One thing that we can do is to determine whether or not there are any numbers in the labels which are not integers (or very very close to integers), and in that case, we can cast the matrix to a `Mat<size_t>` or `Mat<int>` and then save that.  However, we should only do this for the case of text-file matrices (csv, tsv, txt)... otherwise, we might load an `arma_binary` matrix with packed doubles but then save it as packed ints, causing disaster later on in the user's pipeline.

If this is too hard, I don't think that it is the end of the world to turn the user's '1's and '0's into '1.0000000000e+00' and '0.00000000000e+00' since those are still valid number representations and will work later in the user's machine learning pipeline.  But we should try to avoid reimplementing Armadillo's save functionality, whatever it is we do choose to do.

zoq commented:
I agree, that ```ios::scientific``` is the best way to represent arbitrary numbers, but I don't see a reason to use a default precision value. If we could overwrite the precision value, we could read the precision once we load the file, and use the same precision to save the file again.

I agree, that using a precision of 12 is not a big deal when it comes to small files, but I think, it is a problem once you have a 'bigger' file (values with precision(3)) and use the split function.

rcurtin commented:
Another consideration here is that if space is really a concern, csv/tsv/txt is really about the most inefficient possible format for representing doubles.  (The situation is a little bit different for small integers...).  Either using `csv.gz` or HDF5 or any of the packed binary formats is a much better option in that case (and actually accelerates load times by sometimes an order of magnitude or more).

If we are doing something like reading the precision of the file when we load it in order that we can have the same precision when we save it, this is definitely a way to solve our problem, but to me it's not too clear how much extra overhead we'll have when we load files... do we then need to modify `data::Load()` to return a `size_t` (or take a `size_t&` as a parameter) which can then be passed to `data::Save()`?  But even then this `size_t` wouldn't be useful for packed binary formats.

So I am definitely not opposed to a change in how we load things to improve this behavior, but I'm not quite sure what the best way to do it is.  Contributions to Armadillo upstream to change its behavior are definitely something we can do (though Conrad may or may not like the ideas we propose).

stereomatchingkiss commented:
>do we then need to modify data::Load() to return a size_t (or take a size_t& as a parameter) which can then be passed to data::Save()? 

Do you mean save the precision value when saving the file?Without the precision value, I do not know how could we measure the precision of double after we load the file.

I think the easiest solution is ask the users save their big files by hdf5 format(never try csv.gz), this is why something like hdf5 exist from the beginning.

@keonkim Please give me some times to study your codes, play with it and do some test.

stereomatchingkiss commented:
@keonkim  Looks good for me, only something need to refine

1 : format 
Split functions have some trailing spaces, please remove them

2 : move the train and test data into std::make_tuple
The overload of Split function which return tuple, please move the trainData and testData into the tuple.

`return std::make_tuple(std::move(trainData), std::move(testData));`

3 : `if ((testRatio < 0.0) && (testRatio > 1.0))`

should be

`if ((testRatio < 0.0) || (testRatio > 1.0))`

4 :  Following codes will copy the value, you can

```
mat training = get<0>(value);
mat test = get<1>(value);
mat trainingLabels = get<2>(value);
mat testLabels = get<3>(value);
```

a : move it
```
mat training = std::move(get<0>(value));
mat test = std::move(get<1>(value));
mat trainingLabels = std::move(get<2>(value));
mat testLabels = std::move(get<3>(value));

```
b : use tuple directly
```
data::Save(trainingFile, get<0>(value), false);
data::Save(testFile, get<1>(value), false);
data::Save(trainingLabelsFile, get<2>(value), false);
data::Save(testLabelsFile, get<3>(value), false);
```

c : use another api from the beginning
```
mat training;
mat test;
mat trainingLabels;
mat testLabels;
data::Split(data, labels_row, training, test, trainingLabels, testLabels, testRatio);
```
d : take the reference, with const( I love const :) )
```
mat const &training = std::move(get<0>(value));
mat const &test = std::move(get<1>(value));
mat const &trainingLabels = std::move(get<2>(value));
mat const &testLabels = std::move(get<3>(value));

```

split without parameters part has the same problem

5 : The test of non-labels Split haven't finished

The file only test the size after split, but do not test the data after are split into correct clusters(train and test). To test it, I think we could 

step 1 : generate sequential, unique values to the input data
step 2 : split input data into train data and test data
step 3 : make sure all of the data in input data could be found in train data and test data

keonkim commented:
@stereomatchingkiss Thanks for the comment, I will update it :)

stereomatchingkiss commented:
@keonkim Looking forward to that. 

ps : I would pick solution b or c, less cumbersome

zoq commented:
Regarding the space/precision problem, I agree using a packed binary format solves the space problem.

It's just that I wouldn't expect that if I use the tool to split my data, I probably end up with a bigger file. However, I don't think, we should rewrite the load and save.

zoq commented on src/mlpack/methods/preprocess/preprocess_split_main.cpp:
```diff
@@ -0,0 +1,134 @@
+        << testLabelsFile << endl;
+    }
+  }
+  else
```
We could also assume that the last dimension of the input data are the input labels, that's what the executables do.

zoq replied:
I think I get the intention, you like to preserve the data structure. In this case, I take it back.

zoq commented on src/mlpack/tests/split_data_test.cpp:
```diff
@@ -67,17 +83,31 @@ void CheckDuplication(const Row<size_t>& trainLabels,
+{
+  mat input(2, 10);
+  size_t count = 0; // count for putting unique sequential values
+  input.imbue([&count] () { return ++count; });
```
Really neat idea, I propably just used:

  ```arma::mat input = arma::reshape(arma::linspace<arma::mat>(0,  19,  20), 2, 10);```

  but your idea is great.

stereomatchingkiss replied:
I agree, really neat idea

keonkim replied:
thanks :)

stereomatchingkiss commented:
@keonkim Nice job, I think everything are fine.

rcurtin commented on src/mlpack/tests/split_data_test.cpp:
```diff
@@ -41,6 +41,22 @@ void CompareData(const mat& inputData,
+    const mat& rhsCol = sortedCompare.col(i);
+    for (size_t j = 0; j < lhsCol.n_rows; ++j)
+    {
+      BOOST_REQUIRE_CLOSE(lhsCol(j), rhsCol(j), 1e-5);
```
This can fail for very small values, I like to use

```
if (std::abs(rhsCol(j)) < 1e-5)
  BOOST_REQUIRE_SMALL(lhsCol(j), 1e-5);
else
  BOOST_REQUIRE_CLOSE(lhsCol(j), rhsCol(j), 1e-5);
```

This is a nice utility function that we could probably use, though.  Actually now that I am thinking about it more... I wonder what would be necessary for us to be able to write `BOOST_REQUIRE_CLOSE(lhs, rhs, 1e-5)` and get the comparison here?  I dunno if that is worth taking the time to figure out though :)

rcurtin commented:
@stereomatchingkiss I thought you were an admin too, I sent the invite :)

Looks good to me also, I think I would like to write some more documentation in the `PROGRAM_INFO()` macro but I can do that after merge.

keonkim commented:
@rcurtin I can add more docs tomorrow

rcurtin commented:
Up to you, if you want to do it that would be just fine.  Here is an example of what I would be shooting for:

https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/kmeans/kmeans_main.cpp

In this case, the program is simpler, so fewer details are necessary, but basically, in PROGRAM_INFO you should be writing the man page (and actually, you are, because if you type `make man`, then it generates a man page for each program by processing the output of `bin/<program_name> --help`).

stereomatchingkiss commented:
@keonkim Thanks for adding doc and your contributions :).

keonkim commented:
@rcurtin @stereomatchingkiss @zoq
Thanks!
So.. do you think merge now?
